src/dex/generic-rfq/rate-fetcher.ts

```diff
@@ -0,0 +1,387 @@
+    if (!parsed.markets) {
+      return null;
+    }
+    return parsed;
+  }
+
+  private castRateResponse(data: unknown): RatesResponse | null {
+    if (!data) {
+      return null;
+    }
+
+    return data as RatesResponse;
+  }
+
+  start() {
+    this.marketFetcher.startPolling();
+  }
+
+  stop() {
+    this.marketFetcher.stopPolling();
+    this.rateFetcher.stopPolling();
+  }
+
+  private handleMarketResponse(resp: MarketResponse) {
+    this.pairs = {};
+
+    if (this.rateFetcher.isPolling()) {
+      this.rateFetcher.stopPolling();
+    }
+
+    const tokensAddress: string[] = [];
+    const pairsFullName: string[] = [];
+    const pairs: PairMap = {};
+    for (const pair of resp.markets) {
+      if (pair.status !== 'available') {
+        continue;
+      }
+      pairs[pair.id] = {
+        ...pair,
+      };
+      tokensAddress.push(pair.base.address);
+      tokensAddress.push(pair.quote.address);
+
+      const fullName = `${pair.quote.address}_${pair.base.address}`;
+      const reversedFullName = `${pair.base.address}_${pair.quote.address}`;
+      const reversedId = pair.id.split('-').reverse().join('-');
+      const reversedPair = {
+        id: reversedId,
+        fullName: fullName,
+        base: pair.quote,
+        quote: pair.base,
+        status: pair.status,
+      };
+
+      pairs[reversedId] = reversedPair;
+
+      pairsFullName.push(fullName);
+      pairsFullName.push(reversedFullName);
+    }
+
+    this.dexHelper.cache.setex(
+      this.dexKey,
+      this.dexHelper.config.data.network,
+      'tokens',
+      this.config.marketConfig.dataTTLS,
+      JSON.stringify(tokensAddress.filter(onlyUnique)),
+    );
+
+    this.dexHelper.cache.setex(
+      this.dexKey,
+      this.dexHelper.config.data.network,
+      'pairs',
+      this.config.marketConfig.dataTTLS,
+      JSON.stringify(pairsFullName.filter(onlyUnique)),
+    );
+
+    this.pairs = pairs;
+    this.rateFetcher.startPolling();
+  }
+
+  private handleRatesResponse(resp: RatesResponse) {
+    const pairs = this.pairs;
+    for (const [pairName, price] of Object.entries(resp)) {
+      const pair = pairs[pairName];
+      if (!pair) {
+        continue;
+      }
+      const prices = resp[pairName];
+
+      if (prices.bids.length) {
+        this.dexHelper.cache.setex(
+          this.dexKey,
+          this.dexHelper.config.data.network,
+          `${pair.base.address}_${pair.quote.address}_${SwapSide.SELL}`,
+          this.config.marketConfig.dataTTLS,
+          JSON.stringify(prices.bids),
+        );
+      }
+
+      if (prices.asks.length) {
+        this.dexHelper.cache.setex(
+          this.dexKey,
+          this.dexHelper.config.data.network,
+          `${pair.base.address}_${pair.quote.address}_${SwapSide.BUY}`,
+          this.config.marketConfig.dataTTLS,
+          JSON.stringify(prices.asks),
+        );
+      }
+    }
+  }
+
+  checkHealth(): boolean {
+    return [this.marketFetcher, this.rateFetcher].some(
```

Comment:
Maybe better `every`?

Reply:
Currently it's some on hs

Reply:
But what if one is not functioning? We will not know. Normally both of them should work

Reply:
I think we should not fail when a third party is failing. Otherwise it will restart api. Btw currently checkHealth is not call anywhere. Because I would like to build on top of initialization service PR